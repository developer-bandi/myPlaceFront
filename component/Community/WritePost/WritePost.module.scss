@import "../../../lib/styles/globalmixin.scss";

.mainBlock {
  width: 800px;
  margin: 30px auto 30px auto;
  min-height: calc(100vh - 242px);
  @media screen and (max-width: 840px) {
    width: calc(100% - 40px);
    margin: 30px 20px 30px 20px;
  }
  @media screen and (max-width: 767px) {
    min-height: calc(100vh - 302px);
  }
}

.title {
  width: 100%;
  font-size: 35px;
  font-weight: 500;
  margin: 0px 0px 30px 0px;
  @media screen and (max-width: 840px) {
    font-size: 30px;
  }
}

.titleInput {
  width: 100%;
  height: 40px;
  border: 1px solid #d1d5db;
  border-radius: 5px;
  padding: 10px 10px 10px 10px;
  margin: 0px 0px 30px 0px;
  &:focus {
    border: 1px solid black;
  }
  @media screen and (max-width: 500px) {
    width: 100%;
  }
}

.photoBlock {
  display: flex;
  align-items: center;
  margin-bottom: 30px;
  width: 100%;
  padding: 15px 15px 15px 15px;
  border: 1px solid #d1d5db;
  border-radius: 5px;
}

.photoInput {
  width: 800px;
  height: 250px;
}

.contentInput {
  width: 100%;
  height: 250px;
  border-radius: 5px;
  border: 1px solid #d1d5db;
  padding: 20px;
  &:focus {
    border: 1px solid black;
  }
}

.ButtonBlock {
  width: 100%;
  display: flex;
  justify-content: flex-end;
  margin-top: 20px;
}

.submitButton {
  width: 80px;
  height: 40px;
  background: #00aaff;
  color: white;
  border-radius: 5px;
  cursor: pointer;
  &:hover {
    opacity: 0.5;
  }
}

.fileButton {
  border: 1px solid #eaeaea;
  display: block;
  width: 100px;
  height: 100px;
  border-radius: 10px;
  color: gray;
  font-weight: bold;
  font-size: 20px;
  padding-top: 20px;
  margin-right: 10px;
  flex: 0 0 auto;
  cursor: pointer;
}

.realFileButton {
  display: none;
}

.fileIcon {
  width: 40px;
  margin: auto;
}

.fileCount {
  width: 38px;
  margin: auto;
  font-size: 15px;
}

.imgDeleteButton {
  position: absolute;
  background: black;
  opacity: 0.7;
  font-size: 25px;
  width: 35px;
  height: 35px;
  top: 3%;
  right: 3%;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 10px;
  &:hover {
    cursor: pointer;
  }
}

.imgBlock {
  height: 100px;
  margin-right: 10px;
  flex: 0 0 auto;
  position: relative;
}

.img {
  border-radius: 10px;
  margin-right: 10px;
}
