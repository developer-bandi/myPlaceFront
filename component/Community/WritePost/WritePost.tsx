import Image from "next/image";
import {RefObject} from "react";
import {AiOutlinePicture} from "react-icons/ai";
import {ImCross} from "react-icons/im";
import styles from "./WritePost.module.scss";

interface WritePostProps {
  addImg: (e: React.ChangeEvent<HTMLInputElement>) => void;
  deleteImg: (deleteindex: number) => void;
  uploadImg: string[];
  titleRef: RefObject<HTMLInputElement>;
  contentRef: RefObject<HTMLTextAreaElement>;
  submit: () => void;
}

const WritePost = ({
  addImg,
  deleteImg,
  uploadImg,
  titleRef,
  contentRef,
  submit,
}: WritePostProps) => {
  return (
    <main className={styles.mainBlock}>
      <div className={styles.title}>포스트 작성하기</div>
      <input
        className={styles.titleInput}
        ref={titleRef}
        placeholder="제목을 입력하세요"
      ></input>
      <ul className={styles.photoBlock}>
        <li>
          <label htmlFor="fileimg" className={styles.fileButton}>
            <div className={styles.fileIcon}>
              <AiOutlinePicture size="40" />
            </div>
            <div className={styles.fileCount}>{`(${
              uploadImg === undefined ? 0 : uploadImg.length
            }/5)`}</div>
          </label>
        </li>
        {(uploadImg || []).map((img, index) => {
          return (
            <li className={styles.imgBlock} key={index}>
              <Image
                src={img}
                width={"100px"}
                height={"100px"}
                className={styles.img}
              ></Image>
              <div
                className={styles.imgDeleteButton}
                onClick={() => deleteImg(index)}
                data-testid={`deleteImg${index}`}
              >
                <ImCross
                  size={20}
                  style={{color: "white", opacity: "1"}}
                ></ImCross>
              </div>
            </li>
          );
        })}
      </ul>
      <input
        type="file"
        accept="image/*"
        id="fileimg"
        onChange={addImg}
        multiple
        className={styles.realFileButton}
        data-testid="addImg"
      />
      <div className={styles.contentBlock}>
        <textarea
          className={styles.contentInput}
          ref={contentRef}
          placeholder="내용을 입력하세요"
        />
      </div>
      <div className={styles.ButtonBlock}>
        <button
          className={styles.submitButton}
          onClick={submit}
          data-testid="submit"
        >
          등록
        </button>
      </div>
    </main>
  );
};

export default WritePost;
