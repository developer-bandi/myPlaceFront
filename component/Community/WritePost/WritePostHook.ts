import {useRouter} from "next/router";
import {useCallback, useRef, useState} from "react";
import {useSelector} from "react-redux";
import {axiosPostDetail} from "../../../lib/commonFn/api";
import {
  addImgWrapper,
  deleteUploadImgWrapper,
} from "../../../lib/commonFn/imgFn";
import {RootReducer} from "../../../store";

const useWritePost = () => {
  const titleRef = useRef<HTMLInputElement>(null);
  const contentRef = useRef<HTMLTextAreaElement>(null);
  const loginedUser = useSelector(
    (state: RootReducer) => state.userLogin.content
  );
  const router = useRouter();
  const [uploadImg, setUploadImg] = useState<string[]>([]);
  const [imgfile, setImgfile] = useState<Blob[]>([]);

  const addImg = useCallback(
    addImgWrapper(5, uploadImg, imgfile, setUploadImg, setImgfile),
    [uploadImg, imgfile, setUploadImg, setImgfile]
  );

  const deleteImg = useCallback(
    deleteUploadImgWrapper(uploadImg, imgfile, setUploadImg, setImgfile),
    [uploadImg, imgfile, setUploadImg, setImgfile]
  );

  const submit = async () => {
    if (!loginedUser) {
      alert("로그인을 해주세요!");
    } else {
      if (contentRef.current !== null && titleRef.current !== null) {
        try {
          await axiosPostDetail(
            titleRef.current.value,
            contentRef.current.value,
            imgfile
          );
          alert("게시글 등록이 완료되었습니다");
          router.push("/community/postlist");
        } catch (error) {
          alert("에러가 발생하였습니다");
        }
      }
    }
  };

  return {
    addImg,
    deleteImg,
    uploadImg,
    titleRef,
    contentRef,
    submit,
  };
};

export default useWritePost;
