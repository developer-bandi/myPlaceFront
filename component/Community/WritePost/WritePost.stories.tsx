import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import WritePost from "./WritePost";
import { action } from "@storybook/addon-actions";

export default {
  title: "Community/WritePost",
  component: WritePost,
} as ComponentMeta<typeof WritePost>;

const Template: ComponentStory<typeof WritePost> = (arg) => (
  <WritePost {...arg} />
);

export const generalType = Template.bind({});
generalType.args = {
  uploadImg: [],
  addImg: action("addImg"),
  submit: action("submit"),
};
