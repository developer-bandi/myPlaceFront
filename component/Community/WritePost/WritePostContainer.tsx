import WritePost from "./WritePost";
import useWritePost from "./WritePostHook";

const WritePostContainer = () => {
  const {addImg, deleteImg, uploadImg, titleRef, contentRef, submit} =
    useWritePost();
  return (
    <WritePost
      addImg={addImg}
      deleteImg={deleteImg}
      uploadImg={uploadImg}
      titleRef={titleRef}
      contentRef={contentRef}
      submit={submit}
    />
  );
};

export default WritePostContainer;
