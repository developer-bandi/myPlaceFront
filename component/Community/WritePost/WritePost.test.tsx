import {
  act,
  cleanup,
  fireEvent,
  render,
  renderHook,
  screen,
} from "@testing-library/react";
import axios from "axios";
import {useRouter} from "next/router";
import React, {ReactNode} from "react";
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import WritePost from "./WritePost";
import useWritePost from "./WritePostHook";
jest.mock("axios");
jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("writePost Hook test", () => {
  describe("submit 함수 test", () => {
    afterEach(cleanup);
    let loginStatus = false;
    const mockStore = configureMockStore();
    const loginStoreMock = mockStore({
      userLogin: {
        content: {id: 1, nickname: "test"},
        loading: false,
        error: false,
      },
    });
    const logoutStoreMock = configureMockStore()({
      userLogin: {
        loading: false,
        error: false,
      },
    });
    const wrapper = ({children}: {children: ReactNode}) => (
      <Provider store={loginStatus ? loginStoreMock : logoutStoreMock}>
        {children}
      </Provider>
    );

    it("로그인 하지 않은 경우", async () => {
      window.alert = jest.fn();
      jest.spyOn(React, "useRef").mockReturnValueOnce({current: "test"});
      const {result} = renderHook(() => useWritePost(), {
        wrapper,
      });
      await act(async () => {
        await result.current.submit();
      });
      expect(window.alert).toBeCalledWith("로그인을 해주세요!");
    });

    it("정상적으로 업로드 한경우", async () => {
      loginStatus = true;
      window.alert = jest.fn();
      const mockRouter = {
        push: jest.fn(),
      };
      (useRouter as jest.Mock).mockReturnValue(mockRouter);
      jest.spyOn(React, "useRef").mockReturnValue({current: "test"});

      mockedAxios.post.mockImplementation(() =>
        Promise.resolve({status: 200, data: "test"})
      );
      const {result} = renderHook(() => useWritePost(), {
        wrapper,
      });
      await act(async () => {
        await result.current.submit();
      });
      expect(window.alert).toBeCalledWith("게시글 등록이 완료되었습니다");
      expect(mockRouter.push).toBeCalledWith("/community/postlist");
    });

    it("업로드중 에러가 발생한 경우", async () => {
      window.alert = jest.fn();
      const useRefSpy = jest
        .spyOn(React, "useRef")
        .mockReturnValueOnce({current: "test"});
      mockedAxios.post.mockImplementation(() =>
        Promise.reject({status: 500, data: "에러발생"})
      );
      const {result} = renderHook(() => useWritePost(), {
        wrapper,
        initialProps: {
          login: false,
        },
      });
      await act(async () => {
        await result.current.submit();
      });
      expect(window.alert).toBeCalledWith("에러가 발생하였습니다");
    });
  });
});

describe("WritePost Presentational Test", () => {
  const addImgMock = jest.fn();
  const deleteImgMock = jest.fn();
  const submitMock = jest.fn();
  const utils = render(
    <WritePost
      addImg={addImgMock}
      deleteImg={deleteImgMock}
      uploadImg={["http://test"]}
      titleRef={{current: null}}
      contentRef={{current: null}}
      submit={submitMock}
    />
  );
  expect(utils.container).toMatchSnapshot();
  fireEvent.click(screen.getByTestId("deleteImg0"));
  fireEvent.change(screen.getByTestId("addImg"), {
    target: {value: ""},
  });
  fireEvent.click(screen.getByTestId("submit"));

  expect(addImgMock).toBeCalledTimes(1);
  expect(deleteImgMock).toBeCalledTimes(1);
  expect(submitMock).toBeCalledTimes(1);
});
