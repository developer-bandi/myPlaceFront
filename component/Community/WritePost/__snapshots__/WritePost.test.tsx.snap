// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <main
    class="mainBlock"
  >
    <div
      class="title"
    >
      포스트 작성하기
    </div>
    <input
      class="titleInput"
      placeholder="제목을 입력하세요"
    />
    <ul
      class="photoBlock"
    >
      <li>
        <label
          class="fileButton"
          for="fileimg"
        >
          <div
            class="fileIcon"
          >
            <svg
              fill="currentColor"
              height="40"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 632H136v-39.9l138.5-164.3 150.1 178L658.1 489 888 761.6V792zm0-129.8L664.2 396.8c-3.2-3.8-9-3.8-12.2 0L424.6 666.4l-144-170.7c-3.2-3.8-9-3.8-12.2 0L136 652.7V232h752v430.2zM304 456a88 88 0 1 0 0-176 88 88 0 0 0 0 176zm0-116c15.5 0 28 12.5 28 28s-12.5 28-28 28-28-12.5-28-28 12.5-28 28-28z"
              />
            </svg>
          </div>
          <div
            class="fileCount"
          >
            (1/5)
          </div>
        </label>
      </li>
      <li
        class="imgBlock"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27100%27%20height=%27100%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            class="img"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
        <div
          class="imgDeleteButton"
          data-testid="deleteImg0"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            style="color: white; opacity: 1;"
            version="1.1"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.854 12.854c-0-0-0-0-0-0l-4.854-4.854 4.854-4.854c0-0 0-0 0-0 0.052-0.052 0.090-0.113 0.114-0.178 0.066-0.178 0.028-0.386-0.114-0.529l-2.293-2.293c-0.143-0.143-0.351-0.181-0.529-0.114-0.065 0.024-0.126 0.062-0.178 0.114 0 0-0 0-0 0l-4.854 4.854-4.854-4.854c-0-0-0-0-0-0-0.052-0.052-0.113-0.090-0.178-0.114-0.178-0.066-0.386-0.029-0.529 0.114l-2.293 2.293c-0.143 0.143-0.181 0.351-0.114 0.529 0.024 0.065 0.062 0.126 0.114 0.178 0 0 0 0 0 0l4.854 4.854-4.854 4.854c-0 0-0 0-0 0-0.052 0.052-0.090 0.113-0.114 0.178-0.066 0.178-0.029 0.386 0.114 0.529l2.293 2.293c0.143 0.143 0.351 0.181 0.529 0.114 0.065-0.024 0.126-0.062 0.178-0.114 0-0 0-0 0-0l4.854-4.854 4.854 4.854c0 0 0 0 0 0 0.052 0.052 0.113 0.090 0.178 0.114 0.178 0.066 0.386 0.029 0.529-0.114l2.293-2.293c0.143-0.143 0.181-0.351 0.114-0.529-0.024-0.065-0.062-0.126-0.114-0.178z"
            />
          </svg>
        </div>
      </li>
    </ul>
    <input
      accept="image/*"
      class="realFileButton"
      data-testid="addImg"
      id="fileimg"
      multiple=""
      type="file"
    />
    <div
      class="contentBlock"
    >
      <textarea
        class="contentInput"
        placeholder="내용을 입력하세요"
      />
    </div>
    <div
      class="ButtonBlock"
    >
      <button
        class="submitButton"
        data-testid="submit"
      >
        등록
      </button>
    </div>
  </main>
</div>
`;
